src/analyses/variable-sensitivity/abstract_environment.cpp

```diff
@@ -468,3 +499,211 @@ std::vector<abstract_object_pointert> eval_operands(
 
   return operands;
 }
+
+///////////
+abstract_value_pointert as_value(const abstract_object_pointert &obj)
+{
+  auto context_value =
+    std::dynamic_pointer_cast<const context_abstract_objectt>(obj);
+
+  return context_value
+           ? as_value(context_value->unwrap_context())
+           : std::dynamic_pointer_cast<const abstract_value_objectt>(obj);
+}
+
+bool is_value(const abstract_object_pointert &obj)
+{
+  return as_value(obj) != nullptr;
+}
+
+static auto inverse_operations =
+  std::map<irep_idt, irep_idt>{{ID_equal, ID_notequal},
+                               {ID_notequal, ID_equal},
+                               {ID_le, ID_gt},
+                               {ID_lt, ID_ge},
+                               {ID_ge, ID_lt},
+                               {ID_gt, ID_le}};
+
+exprt invert_expr(const exprt &expr)
+{
+  auto expr_id = expr.id();
+  auto inverse_operation = inverse_operations.find(expr_id);
+  if(inverse_operation == inverse_operations.end())
+    return nil_exprt();
```

Comment:
Having rewritten `assume_or` in terms of `assume_and`, I think this is taken care of. If we're assuming `not_exprt(or_exprt(A, B))` then that will be transformed to (more or less) `not_expr(not_expr(and_exprt(not_exprt(A), not_exprt(B))))`. It's a little long winded, but I think it will evaluate correctly.

Reply:
Sounds good; thanks.